sdk/ai/azure-ai-inference/src/main/java/com/azure/ai/inference/ChatCompletionClientTracer.java

```diff
@@ -0,0 +1,606 @@
+        final CompletionsUsage usage = response.getUsage();
+        if (usage != null) {
+            tracer.setAttribute("gen_ai.usage.input_tokens", usage.getPromptTokens(), span);
+            tracer.setAttribute("gen_ai.usage.output_tokens", usage.getCompletionTokens(), span);
+        }
+        tracer.setAttribute("gen_ai.response.finish_reasons", finishReasons, span);
+    }
+
+    private void traceCompletionResponseEvents(ChatCompletions response, Context span) {
+        final List<ChatChoice> choices = response.getChoices();
+        if (choices != null) {
+            final OffsetDateTime now = OffsetDateTime.now(ZoneOffset.UTC);
+            for (ChatChoice choice : choices) {
+                traceChoiceEvent(toJsonString(choice), now, span);
+            }
+        }
+    }
+
+    private void traceChoiceEvent(String choiceContent, OffsetDateTime timestamp, Context span) {
+        final Map<String, Object> eventAttributes = new HashMap<>(2);
+        eventAttributes.put("gen_ai.system", INFERENCE_GEN_AI_SYSTEM_NAME);
+        eventAttributes.put("gen_ai.event.content", choiceContent);
+        tracer.addEvent("gen_ai.choice", eventAttributes, timestamp, span);
+    }
+
+    private String toJsonString(ChatRequestMessage message) {
+        try (ByteArrayOutputStream stream = new ByteArrayOutputStream();
+            JsonWriter writer = JsonProviders.createWriter(stream)) {
+            message.toJson(writer);
+            writer.flush();
+            return new String(stream.toByteArray(), StandardCharsets.UTF_8);
+        } catch (IOException e) {
+            logger.atWarning().log("'ChatRequestMessage' serialization error", e);
+        }
+        return null;
+    }
+
+    private String toJsonString(ChatChoice choice) {
+        try (ByteArrayOutputStream stream = new ByteArrayOutputStream();
+            JsonWriter writer = JsonProviders.createWriter(stream)) {
+            writer.writeStartObject();
+            writer.writeStartObject("message");
+            if (captureContent) {
+                writer.writeStringField("content", choice.getMessage().getContent());
+            }
+            if (choice.getMessage() != null) {
+                final List<ChatCompletionsToolCall> toolCalls = choice.getMessage().getToolCalls();
+                if (toolCalls != null && !toolCalls.isEmpty()) {
+                    writer.writeStartArray("tool_calls");
+                    for (ChatCompletionsToolCall toolCall : toolCalls) {
+                        if (captureContent) {
+                            toolCall.toJson(writer);
+                        } else {
+                            writer.writeStartObject();
+                            writer.writeStringField("id", toolCall.getId());
+                            writer.writeStringField("type", toolCall.getType());
+                            writer.writeEndObject();
+                        }
+                    }
+                    writer.writeEndArray();
```

Comment:
You could leverage `JsonWriter.writeArrayField(String, List<T>, WriteValueCallback<JsonWriter, T>)` here to simplify the code a bit.

```suggestion
                    writer.writeArrayField("tool_calls", toolCalls, (w, toolCall) -> {
                        if (captureContent) {
                            toolCall.toJson(w);
                        } else {
                            w.writeStartObject();
                            w.writeStringField("id", toolCall.getId());
                            w.writeStringField("type", toolCall.getType());
                            w.writeEndObject();
                        }
                    });
```

Reply:
updated to use `   writer.writeArrayField`